src/terminator/mod.rs

```diff
@@ -559,6 +559,36 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
         let usize = self.tcx.types.usize;
 
         match &link_name[..] {
+            "malloc" => {
+                let size = self.value_to_primval(args[0], usize)?.to_u64()?;
+                let align = self.memory.pointer_size();
+                let ptr = self.memory.allocate(size, align)?;
```

Comment:
> The allocate function handles zst allocs right now

Yes, and it permits them -- whereas `__rust_allocate` and friends really should fail for size 0.